include/lbann/callbacks/checkpoint.hpp

```diff
@@ -151,47 +170,67 @@ class checkpoint : public callback_base {
   };
 };
 
-inline std::string get_last_shared_checkpoint_filename(model *m, const std::string& dir) {
-  lbann_comm *comm = m->get_comm();
+inline std::string get_trainer_checkpoint_dirname(const std::string& trainer_name, const std::string& dir) {
   std::ostringstream ss;
   ss << dir << "/";
-  ss << m->get_name().c_str() << ".trainer.";
-  ss << comm->get_trainer_rank() << ".last.shared.checkpoint";
+  ss << trainer_name.c_str() << "/";
+  return ss.str();
+}
+
+inline std::string get_last_shared_checkpoint_filename(model *m, const std::string& dir) {
+  std::ostringstream ss;
+  ss << dir;
+  ss << m->get_name().c_str();
+  ss << ".last.shared.checkpoint";
   return ss.str();
 }
 
+inline std::string get_last_shared_checkpoint_filename(const std::string& trainer_name, model *m, const std::string& dir) {
+  return get_last_shared_checkpoint_filename(m, get_trainer_checkpoint_dirname(trainer_name, dir));
+}
+
 inline std::string get_shared_checkpoint_dirname(model *m, const std::string& dir, execution_mode mode, size_t epoch, size_t step) {
-  lbann_comm *comm = m->get_comm();
   std::ostringstream ss;
-  ss << dir << "/" << m->get_name().c_str();
-  ss << ".trainer." << comm->get_trainer_rank();
+  ss << dir;
+  ss << m->get_name().c_str();
   ss << ".shared." << to_string(mode);
   ss << ".epoch." << epoch;
   ss << ".step."<< step << "/";
   return ss.str();
 }
 
+inline std::string get_shared_checkpoint_dirname(const std::string& trainer_name, model *m, const std::string& dir, execution_mode mode, size_t epoch, size_t step) {
+  return get_shared_checkpoint_dirname(m, get_trainer_checkpoint_dirname(trainer_name, dir), mode, epoch, step);
+}
+
 inline std::string get_last_distributed_checkpoint_filename(model *m, const std::string& dir) {
-  lbann_comm *comm = m->get_comm();
   std::ostringstream ss;
-  ss << dir << "/";
-  ss << m->get_name().c_str() << ".trainer.";
-  ss << comm->get_trainer_rank() << ".last.distributed.checkpoint";
+  ss << dir;
+  ss << m->get_name().c_str();
+  ss << ".last.distributed.checkpoint";
   return ss.str();
 }
 
+inline std::string get_last_distributed_checkpoint_filename(const std::string& trainer_name, model *m, const std::string& dir) {
+  return get_last_distributed_checkpoint_filename(m, get_trainer_checkpoint_dirname(trainer_name, dir));
+}
+
 inline std::string get_distributed_checkpoint_dirname(model *m, const std::string& dir, execution_mode mode, size_t epoch, size_t step) {
   lbann_comm *comm = m->get_comm();
   std::ostringstream ss;
-  ss << dir << "/" << m->get_name().c_str();
-  ss << ".trainer." << comm->get_trainer_rank();
+  ss << dir;
+  ss << m->get_name().c_str();
   ss << ".rank." << comm->get_rank_in_trainer();
   ss << ".distributed." << to_string(mode);
   ss << ".epoch." << epoch;
   ss << ".step."<< step << "/";
   return ss.str();
```

Comment:
```suggestion
  return build_string(dir,
    m->get_name(),
    ".rank.", comm->get_rank_in_trainer(),
    ".distributed.", to_string(mode),
    ".epoch.", epoch,
    ".step.", step, '/');
```

Reply:
Do you need the "/" after `dir` still? (Applies to all of these changes.)

Reply:
yes we do need the trailing slash.